### ISSUE TITLE
[Transactions] Are they really supported by the binary protocol?

### ISSUE THREAD

**Issue by StarpTech** [orientechnologies/orientdb#4045]
Example 1:

```sql
begin
LET t0 = select expand(in) from E where out.Mid = '553e97a7588219501f5af559' and @class = 'BroughtIntoLeft'
LET t1 = delete edge BroughtIntoLeft where out.Mid = '553e97a7588219501f5af559'
LET t2 = create vertex Node cluster DE_business content {"Email":"JohnDoe@local","Mid":"GF456FH6GHGRFGHGF34567","Points":0,"Position":"","DisplayName":"John Doe","Active":true}
LET t3 = create edge BroughtIntoLeft cluster DE_broughtIntoLeft from (select from Node where Mid = '123456789') to $t2
LET t4 = create edge BroughtIntoLeft cluster DE_broughtIntoLeft from $t2 to $t0
commit
return [$t0,$t1,$t2,$t3,$t4]
```

At command `t3` I use a fictive ID `123456789` but it throws no exception or something like that.

Example 2
```sql
begin
LET t0 = select expand(in) from E where out.Mid = '553e96e0f6787a7c11c7b62f' and @class = 'BroughtIntoLeft'
LET t1 = delete edge BroughtIntoLeft where out.Mid = '553e96e0f6787a7c11c7b62f'
LET t2 = create vertex Node cluster DE_business content {"Email":"JohnDoe@local","Mid":"GF456FH6GHGRFGHGF34567","Points":0,"Position":"","DisplayName":"John Doe","Active":true}
LET t3 = create edge BroughtIntoLeft cluster DE_broughtIntoLeft from (select from Node where Mid = '553e96e0f6787a7c11c7b62f') to $t2
LET t4 = create edge BroughtIntoLeft cluster DE_broughtIntoLeft from $t2 to $t8
commit
return [$t0,$t1,$t2,$t3,$t4]
```

At command `t4` I use a transaction result `$t8` which isn't defined in the query. Like expected it throws an exception but my data aren't restored.

Example 3
Another simpler example: This transaction should throw an exception because the Node with the unique index property `Email` `julia@local` is already used.

```sql
begin
LET t0 = create vertex Node cluster DE_business content {"Email":"JohnDoe@local","Mid":"GF456FH6GHGRFGHGF34567","Points":0,"Position":"","DisplayName":"John Doe","Active":true}
LET t1 = create vertex Node cluster DE_business content {"Email":"julia@local","Mid":"FGZTFGZ456733333","Points":0,"Position":"","DisplayName":"Bernd Eck","Active":true}
LET t2 = create edge BroughtIntoLeft cluster DE_broughtIntoLeft from $t0 to $t1
commit
return [$t0,$t1,$t2]
```

My result:
```
[ { '@type': 'd',
    '@class': 'BroughtIntoLeft',
    '': null,
    '@rid': { cluster: 17, position: 7 } } ]
```
No data is written but I expect an `found duplicated key` exception.

*Orientdb: 2.0.8*
*Nodejs Oriento driver*

**The transaction safety is very limited!** Please improve this.

**Comment by AlexFiverr** [comment]
:+1:  encountered all of this scenario in the REST API as well....

**Comment by tglman** [comment]
hi @StarpTech 

I tried to reproduce this problems, and yes i could reproduce some of them, for 'example 2' a fix is merged in development, for 'example 3' we have already test case that cover that case and it works so i wasn't enable to reproduce it, for 'example 1' still doing some investigation.

**Comment by StarpTech** [comment]
Hi @tglman, in your test for example 3 you don't create a unique dataset before you execute your batch script. Does it make a difference?

**Comment by tglman** [comment]
@StarpTech i was actually adding that test now :)

**Comment by StarpTech** [comment]
@tglman awesome :+1: ^^

**Comment by tglman** [comment]
I fixed all the cases you can review it here : https://github.com/orientechnologies/orientdb/blob/develop/graphdb/src/test/java/com/orientechnologies/orient/graph/TestGraphTransactionOnBatch.java 

there is a particular case though : https://github.com/orientechnologies/orientdb/blob/develop/graphdb/src/test/java/com/orientechnologies/orient/graph/TestGraphTransactionOnBatch.java#L165

closing this.

**Comment by StarpTech** [comment]
Is it possible to make a hotfix? I think this is a big security risk when someone trusts in transactions.

**Comment by tglman** [comment]
ported the fixes also to 2.0.x branch, will be out with the 2.0.9

**Comment by StarpTech** [comment]
:+1:  thanks.
